controllers/argocd/role.go

```diff
@@ -279,59 +279,88 @@ func (r *ReconcileArgoCD) reconcileRoleForApplicationSourceNamespaces(name strin
 	return nil
 }
 
-func (r *ReconcileArgoCD) reconcileClusterRole(name string, policyRules []v1.PolicyRule, cr *argoproj.ArgoCD) (*v1.ClusterRole, error) {
+func (r *ReconcileArgoCD) reconcileClusterRole(componentName string, policyRules []v1.PolicyRule, cr *argoproj.ArgoCD) (*v1.ClusterRole, error) {
 
 	allowed := false
 	if allowedNamespace(cr.Namespace, os.Getenv("ARGOCD_CLUSTER_CONFIG_NAMESPACES")) {
+		// namespace is allowed to host cluster-scoped Argo CD instance
 		allowed = true
 	}
 
-	// Check if it is cluster-scoped instance namespace and user doesn't want to use default ClusterRole
-	if allowed && cr.Spec.DefaultClusterScopedRoleDisabled {
-
-		// In case DefaultClusterScopedRoleDisabled was false earlier and default ClusterRole was created, then delete it.
-		existingClusterRole := &v1.ClusterRole{}
-		if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: GenerateUniqueResourceName(name, cr)}, existingClusterRole); err == nil {
+	if allowed && cr.Spec.DefaultClusterScopedRoleDisabled && cr.Spec.AggregatedClusterRolesEnabled {
+		return nil, fmt.Errorf("Custom Cluster Roles and Aggregated Cluster Roles can not be used together.")
+	}
 
-			// Default ClusterRole exists, now delete it
-			if err := r.Client.Delete(context.TODO(), existingClusterRole); err != nil {
-				return nil, fmt.Errorf("failed to delete existing cluster role for the service account associated with %s : %s", name, err)
-			}
+	// if custom ClusterRole mode is enabled then do nothing and return
+	useCustomClusterRole, err := checkCustomClusterRoleMode(r, cr, componentName, allowed)
+	if useCustomClusterRole {
+		if err != nil {
+			return nil, err
 		}
-
-		// Don't create a default ClusterRole
 		return nil, nil
 	}
 
-	clusterRole := newClusterRole(name, policyRules, cr)
-	if err := applyReconcilerHook(cr, clusterRole, ""); err != nil {
-		return nil, err
+	expectedClusterRole := newClusterRole(componentName, policyRules, cr)
+
+	if allowed && cr.Spec.AggregatedClusterRolesEnabled {
+		// if aggregated ClusterRole mode is enabled, then add required fields in ClusterRole
+		configureAggregatedClusterRole(cr, expectedClusterRole, componentName)
+	} else {
+		// if current mode is default mode, but last one was aggregated mode, then delete ClusterRoles for View and Admin permissions
+		if componentName == common.ArgoCDApplicationControllerComponentView || componentName == common.ArgoCDApplicationControllerComponentAdmin {
+			if err := r.Client.Get(context.TODO(), types.NamespacedName{Name: expectedClusterRole.Name}, expectedClusterRole); err == nil {
+				if err := r.Client.Delete(context.TODO(), expectedClusterRole); err != nil {
+					return nil, fmt.Errorf("failed to delete aggregated ClusterRoles: %s", expectedClusterRole.Name)
+				}
+				return nil, nil
+			}
```

Comment:
If err is non-nil, and error is NOT a 'resource not found' error (which we can check using apierrors.IsNotFound), we should return it.